src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1134,6 +1134,47 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
             return null;
         }
 
+#nullable enable
+        internal UnmanagedCallersOnlyAttributeData? TryGetUnmanagedCallersOnlyAttribute(
+            EntityHandle token,
+            IAttributeNamedArgumentDecoder attributeArgumentDecoder,
+            Func<string, TypedConstant, (bool IsCallConvs, ImmutableHashSet<INamedTypeSymbolInternal>? CallConvs, bool IsValid)> unmanagedCallersOnlyDecoder)
+        {
+            AttributeInfo info = FindTargetAttribute(token, AttributeDescription.UnmanagedCallersOnlyAttribute);
+            if (!info.HasValue || info.SignatureIndex != 0 || !TryGetAttributeReader(info.Handle, out BlobReader sigReader))
+            {
+                return null;
+            }
+
+            var unmanagedConventionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+            bool isValid = true;
+
+            if (sigReader.RemainingBytes > 0)
+            {
+                try
+                {
+                    var numNamed = sigReader.ReadUInt16();
+                    for (int i = 0; i < numNamed; i++)
+                    {
+                        var ((name, value), _, _) = attributeArgumentDecoder.DecodeCustomAttributeNamedArgumentOrThrow(ref sigReader);
+                        var namedArgumentDecoded = unmanagedCallersOnlyDecoder(name, value);
+                        if (namedArgumentDecoded.IsCallConvs)
+                        {
+                            isValid = isValid && namedArgumentDecoded.IsValid;
+                            unmanagedConventionTypes = namedArgumentDecoded.CallConvs;
+                        }
+                    }
+                }
+                catch (Exception ex) when (ex is BadImageFormatException or UnsupportedSignatureContent)
+                {
+                    return UnmanagedCallersOnlyAttributeData.Create(ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false);
```

Comment:
>return UnmanagedCallersOnlyAttributeData.Create(ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false) [](start = 20, length = 113)

It looks like we simply ignore presence of other well-known attributes if we cannot crack them due to an invalid metadata.

Reply:
I'm not sure that's true. At least when cracking `Obsolete`, we're inconsistent. We appear to have this behavior:
* If the constructor is either `public ObsoleteAttribute(string? message)` or `public ObsoleteAttribute(string? message, bool error)` and there is bad metadata in the constructor itself, we ignore the attribute entirely.
* If the constructor is perfectly fine, and we encounter bad metadata cracking named properties, we treat the attribute as having whatever the values it has from the constructor, and we default the DiagnosticId and UrlFormat.

IMO, it's better to replicate this second behavior with this attribute, rather than the first behavior.

---
In reply to: [493012289](https://github.com/dotnet/roslyn/pull/47950#discussion_r493012289) [](ancestors = 493012289)

Reply:
> IMO, it's better to replicate this second behavior with this attribute, rather than the first behavior.

A well-known attribute is matched by a constructor signature, if we cannot crack the signature, or it doesn't match something we recognize, we should ignore the attribute.

---
In reply to: [493892346](https://github.com/dotnet/roslyn/pull/47950#discussion_r493892346) [](ancestors = 493892346,493012289)

Reply:
> A well-known attribute is matched by a constructor signature

If we're in this part of decoding, we matched the constructor signature already, and are going through property/field assignments. If we didn't match to the known constructor signature, we return `null` above, ignoring the attribute.

---
In reply to: [493916803](https://github.com/dotnet/roslyn/pull/47950#discussion_r493916803) [](ancestors = 493916803,493892346,493012289)